core/src/main/java/org/apache/calcite/rel/RelCollationImpl.java

```diff
@@ -97,9 +96,11 @@ public boolean isTop() {
   }
 
   public int compareTo(@Nonnull RelMultipleTrait o) {
-    final RelCollationImpl that = (RelCollationImpl) o;
-    final UnmodifiableIterator<RelFieldCollation> iterator =
-        that.fieldCollations.iterator();
+    if (this == o) {
+      return 0;
+    }
+    final RelCollation that = (RelCollation) o;
```

Comment:
Allowing comparisons between instances of different classes seems a bit risky. If somebody creates a new class `RelCollationImpl2` and defines another method `compareTo` then it is not easy/possible to guarantee that the comparison between `x instanceof RelCollationImpl` and `y instanceof RelCollationImpl2` (i.e., `sgn(x.compareTo(y)) ==-sgn(y.compareTo(x)`) is for instance symmetric. Can you elaborate why is it necessary to support this behavior?

If we want to allow comparisons between instances of different classes (under the same interface `RelCollation`) then it is more appropriate to make this method default and move it to the interface with adequate documentation. 

The same comment I think applies for the `satisfies` method.

Reply:
@zabetak, thanks for your comment. 
There is no particular need for this change, I was just looking at this code and though that it might be "more correct" to work at interface level in these methods.

I agree that comparison between classes can be risky, although we already do something like this in `RelDistributionImpl#compareTo`. But I have no problem to remove this change, your example makes clear that it can lead to risky consequences.

I docthink that the change in `satisfies` is harmless. We could move it as default implementation of the `RelCollation` interface, but then I believe it would be inconsistent with other methods in implemented in `RelCollationImpl` that could be moved as default is `RelCollation` too. IMHO `RelCollationImpl` is like our de-facto default implementation of `RelCollation`, so we should put the methods implementations in there.

This is a minor, cosmetic modification, so no big deal if we decide to discard some / all changes.

Reply:
Agree with @zabetak .

Reply:
`RelCollationImpl that = (RelCollationImpl) o;` the original code is also error prone, why we can assume that the passed in instance is always a `RelCollationImpl` ?

Reply:
Then why do we need the `RelCollation` interface? Is it an over-design? Should `RelCollation` be just a concrete class?

Reply:
I guess my comment is also a bit overthinking :) Do as you see fit @rubenada ;)